File: app/src/main/java/com/boardgamegeek/model/GeekListItem.java
package com.boardgamegeek.model;

import android.text.TextUtils;

import com.boardgamegeek.R;
import com.boardgamegeek.provider.BggContract;
import com.boardgamegeek.util.DateTimeUtils;
import com.boardgamegeek.util.StringUtils;

import org.simpleframework.xml.Attribute;
import org.simpleframework.xml.Element;
import org.simpleframework.xml.ElementList;
import org.simpleframework.xml.Root;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Locale;

@Root(name = "item")
public class GeekListItem {
	public static final int INVALID_OBJECT_TYPE_RES_ID = 0;
	private static final SimpleDateFormat FORMAT = new SimpleDateFormat("EEE, dd MMM yyyy HH:mm:ss Z", Locale.US);

	private long postDateTime = DateTimeUtils.UNPARSED_DATE;
	private long editDateTime = DateTimeUtils.UNPARSED_DATE;

	@Attribute private String id;
	@Attribute private String imageid;
	@Attribute private String objectid;
	@Attribute private String objectname;
	@Attribute private String objecttype;
	@Attribute private String postdate;
	@Attribute private String editdate;
	@Attribute private String subtype;
	@Attribute private String thumbs;
	@Attribute private String username;
	@Element(required = false) private String body;
	@ElementList(name = "comment", inline = true, required = false) private ArrayList<GeekListComment> comments;

	public int imageId() {
		return Integer.valueOf(imageid);
	}

	public int getObjectId() {
		return StringUtils.parseInt(objectid, BggContract.INVALID_ID);
	}

	public String getObjectName() {
		return objectname;
	}

	public int getObjectTypeResId() {
		if (objecttype == null) return INVALID_OBJECT_TYPE_RES_ID;

		switch (objecttype) {
			case "thing":
				if ("boardgame".equals(subtype)) {
					return R.string.title_board_game;
				} else if ("boardgameaccessory".equals(subtype)) {
					return R.string.title_board_game_accessory;
				}
				return R.string.title_thing;
			case "company":
				if ("boardgamepublisher".equals(subtype)) {
					return R.string.title_board_game_publisher;
				}
				return R.string.title_company;
			case "person":
				if ("boardgamedesigner".equals(subtype)) {
					return R.string.title_board_game_designer;
				}
				return R.string.title_person;
			case "family":
				if ("boardgamefamily".equals(subtype)) {
					return R.string.title_board_game_family;
				}
				return R.string.title_family;
			case "filepage":
				return R.string.title_file;
			case "geeklist":
				return R.string.title_geeklist;
		}
		return INVALID_OBJECT_TYPE_RES_ID;
	}

	public boolean isBoardGame() {
		return "thing".equals(objecttype);
	}

	public String getObjectUrl() {
		return "https://www.boardgamegeek.com/" + (TextUtils.isEmpty(subtype) ? objecttype : subtype) + "/" + objectid;
	}

	public long getPostDate() {
		postDateTime = DateTimeUtils.tryParseDate(postDateTime, postdate, FORMAT);
		return postDateTime;
	}

	public long getEditDate() {
		editDateTime = DateTimeUtils.tryParseDate(editDateTime, editdate, FORMAT);
		return editDateTime;
	}

	public int getThumbCount() {
		return StringUtils.parseInt(thumbs, 0);
	}

	public ArrayList<GeekListComment> getComments() {
		return comments;
	}

	public String getUsername() {
		return username;
	}

	public String getBody() {
		return body;
	}
}

File: app/src/main/java/com/boardgamegeek/model/GeekListItem.kt
package com.boardgamegeek.model

import android.os.Parcelable
import com.boardgamegeek.provider.BggContract
import kotlinx.parcelize.IgnoredOnParcel
import kotlinx.parcelize.Parcelize

@Parcelize
data class GeekListItem(
    val id: Long = BggContract.INVALID_ID.toLong(),
    val objectId: Int = BggContract.INVALID_ID,
    val objectName: String = "",
    val objectType: ObjectType = ObjectType.Unknown,
    val imageId: Int = 0,
    val username: String = "",
    val body: String = "",
    val numberOfThumbs: Int = 0,
    val postDateTime: Long = 0L,
    val editDateTime: Long = 0L,
    val comments: List<GeekListComment> = emptyList(),
    val thumbnailUrls: List<String>? = null,
    val heroImageUrls: List<String>? = null,
) : Parcelable {
    enum class ObjectType(val type: String, val subtype: String) {
        BoardGame(TYPE_THING, "boardgame"),
        BoardGameAccessory(TYPE_THING, "boardgameaccessory"),
        Thing(TYPE_THING, ""),
        Publisher(TYPE_COMPANY, "boardgamepublisher"),
        Company(TYPE_COMPANY, ""),
        Designer(TYPE_PERSON, "boardgamedesigner"),
        Person(TYPE_PERSON, ""),
        BoardGameFamily(TYPE_FAMILY, "boardgamefamily"),
        Family(TYPE_FAMILY, ""),
        File(TYPE_FILE, ""),
        GeekList(TYPE_GEEKLIST, ""),
        Unknown("", ""),
    }

    @IgnoredOnParcel
    val isBoardGame: Boolean = objectType in listOf(ObjectType.BoardGame, ObjectType.BoardGameAccessory)

    @IgnoredOnParcel
    val objectUrl: String = when {
        objectType.subtype.isNotBlank() -> "https://www.boardgamegeek.com/${objectType.subtype}/$objectId"
        objectType.type.isNotBlank() -> "https://www.boardgamegeek.com/${objectType.type}/$objectId"
        else -> ""
    }

    companion object {
        private const val TYPE_THING = "thing"
        private const val TYPE_COMPANY = "company"
        private const val TYPE_PERSON = "person"
        private const val TYPE_FAMILY = "family"
        private const val TYPE_FILE = "filepage"
        private const val TYPE_GEEKLIST = "geeklist"
    }
}
